# <center> DSC 320 - Week 6 - Exercise 6
## Alana D'Agostino
### Professor Werner

# <center> Data Scaling and Working with Dataframes

# 1. Data Normalization

In [92]:
# No need to import package here; min() and max() are already inside Python

# Create user defined function for data normalization
## Inputs an array with [] (for lists)
def Normalization_data(data_list):
    results = []
    # Assign objects for variables in formula
    min_value = min(data_list)
    max_value = max(data_list)
    denominator = max_value - min_value
    
    # Use 'for' loop because we are working with a list
    ## Create the actual calculation for normalization
    for x in data_list:
        new_x = (x - min_value) / denominator
        results.append(new_x)
        
    # Use 'return' so Python knows you are performing a calculation
    return results

In [93]:
# Test no.1 function

test_list = [2, 4, 7, 9, 12, 15, 26, 39, 70]

test_results = Normalization_data(test_list)
print(test_results)

[0.0, 0.029411764705882353, 0.07352941176470588, 0.10294117647058823, 0.14705882352941177, 0.19117647058823528, 0.35294117647058826, 0.5441176470588235, 1.0]


# 2. Data Standardization

In [94]:
# NOTES:
## z-scores (statistics): mean divided by the standard deviation
## x-bar is sample mean
## s is standard deviation

# Import necessary libraries
## 'Statistics' to perform statistical calculations (standard deviation)
## NumPy for working with arrays
import statistics
import numpy as np

# Create user defined function for data standardization
## Assign objects to variables in formula
def Standardized_data(listed_data):
    new_results = []
    x_bar = statistics.mean(listed_data)
    stnd_dev = np.std(listed_data)
    
    # 'for' loop again
    ## Write the calculation for standardization
    for x in listed_data:
        new_x = (x - x_bar) / stnd_dev
        new_results.append(new_x)
        
    # 'return' keyword again    
    return new_results

In [95]:
# Test no.2 function

list_for_test = [5.1, 7.2, 9.3, 12.4, 15.5, 89, 60, 45, 84, 23, 26, 36]

test_2_results = Standardized_data(list_for_test)
print(test_2_results)

[-1.0414664414556543, -0.966758344886675, -0.8920502483176954, -0.7817668676682497, -0.6714834870188037, 1.943299892895478, 0.911616654561952, 0.3779873933549557, 1.7654234724931461, -0.40466885641530553, -0.29794300417390623, 0.05780983663075793]


In [96]:
# Test using previous dataframe

test_again = Standardized_data(test_list)
print(test_again)

[-0.8912888176441823, -0.7946430422369819, -0.6496743791261811, -0.5530286037189805, -0.40805994060817985, -0.2630912774973791, 0.2684604872422237, 0.896658027389027, 2.3946675462006346]


# 3. Working with a Dataframe

In [97]:
# Import libraries
## Pandas to perform data manipulation
import pandas as pd

# Provide dataset file location for Python
file_path = r'C:\Users\alana\OneDrive\Desktop\DSC 320\calif_housing_data.csv'
df = pd.read_csv(file_path)
# NOTE: The .csv file is already formatted as a dataframe

# Display dataframe on console
df

housing_median_age  total_bedrooms  households  median_income  \
0                      41           129.0         126         8.3252   
1                      21          1106.0        1138         8.3014   
2                      52           190.0         177         7.2574   
3                      52           235.0         219         5.6431   
4                      52           280.0         259         3.8462   
...                   ...             ...         ...            ...   
20635                  25           374.0         330         1.5603   
20636                  18           150.0         114         2.5568   
20637                  17           485.0         433         1.7000   
20638                  18           409.0         349         1.8672   
20639                  16           616.0         530         2.3886   

       median_house_value  
0                452600.0  
1                358500.0  
2                352100.0  
3                341300.0  
4                342200.0  
...                   ...  
20635             78100.0  
20636             77100.0  
20637             92300.0  
20638             84700.0  
20639             89400.0  

[20640 rows x 5 columns]

In [100]:
# Modify dataframe to perform standard deviation calculation
## This will correct error when adding the 'bedrooms_per_household' column in part 3.e.
## Removes rows that contain empty cells (NULL values)
df = df.dropna()

# Print modified dataframe shape
df.shape

(20433, 6)

# a.

> The original dataframe has 20,640 rows.

# b.

> The target vector for this model is 'median_house_value,' which is the middle value or price of the home. This is the value that the model is trying to predict.

# c.

In [99]:
# Working with 'total_bedrooms' and 'households' columns
# Create a new feature with these columns

# Create new feature column
df['bedrooms_per_household'] = df['total_bedrooms'] / df['households']
df['bedrooms_per_household']

0        1.023810
1        0.971880
2        1.073446
3        1.073059
4        1.081081
           ...   
20635    1.133333
20636    1.315789
20637    1.120092
20638    1.171920
20639    1.162264
Name: bedrooms_per_household, Length: 20640, dtype: float64

In [86]:
# Add new feature to dataframe

df

housing_median_age  total_bedrooms  households  median_income  \
0                      41           129.0         126         8.3252   
1                      21          1106.0        1138         8.3014   
2                      52           190.0         177         7.2574   
3                      52           235.0         219         5.6431   
4                      52           280.0         259         3.8462   
...                   ...             ...         ...            ...   
20635                  25           374.0         330         1.5603   
20636                  18           150.0         114         2.5568   
20637                  17           485.0         433         1.7000   
20638                  18           409.0         349         1.8672   
20639                  16           616.0         530         2.3886   

       median_house_value  bedrooms_per_household  
0                452600.0                1.023810  
1                358500.0                0.971880  
2                352100.0                1.073446  
3                341300.0                1.073059  
4                342200.0                1.081081  
...                   ...                     ...  
20635             78100.0                1.133333  
20636             77100.0                1.315789  
20637             92300.0                1.120092  
20638             84700.0                1.171920  
20639             89400.0                1.162264  

[20433 rows x 6 columns]

> This new feature represents the mean number of bedrooms per household. This is a useful value because it is a major factor in the prediction of the value or price of a home. Additionally, this could offer implications about the average square footage of homes in the area, which is another major factor in the value of a home.

# d.

In [87]:
# Working with 'housing_median_age', 'median_income', and 'bedrooms_per_household' columns

# Create new dataframe with columns specified in above comment
new_df = pd.DataFrame(df, columns = ['housing_median_age', 'median_income', 'bedrooms_per_household'])
new_df

housing_median_age  median_income  bedrooms_per_household
0                      41         8.3252                1.023810
1                      21         8.3014                0.971880
2                      52         7.2574                1.073446
3                      52         5.6431                1.073059
4                      52         3.8462                1.081081
...                   ...            ...                     ...
20635                  25         1.5603                1.133333
20636                  18         2.5568                1.315789
20637                  17         1.7000                1.120092
20638                  18         1.8672                1.171920
20639                  16         2.3886                1.162264

[20433 rows x 3 columns]

# e.

In [89]:
# Convert dataframe columns to list format so Python can add them to new dataframe
## Use tolist() function
list_housing_median_age = new_df['housing_median_age'].tolist()
new_list_housing_median_age = Standardized_data(list_housing_median_age)

list_median_income = new_df['median_income'].tolist()
new_list_median_income = Standardized_data(list_median_income)

list_bedrooms_per_household = new_df['bedrooms_per_household'].tolist()
new_list_bedrooms_per_household = Standardized_data(list_bedrooms_per_household)


In [101]:
# Create new dataframe with columns of standardized data
part_e_df = pd.DataFrame(
    {'standardized_housing_median_age': new_list_housing_median_age,
     'standardized_median_income': new_list_median_income,
     'standardized_bedrooms_per_household': new_list_bedrooms_per_household})

In [102]:
# Display dataframe that includes standardized data on console
part_e_df

standardized_housing_median_age  standardized_median_income  \
0                             0.982163                    2.345163   
1                            -0.606210                    2.332632   
2                             1.855769                    1.782939   
3                             1.855769                    0.932970   
4                             1.855769                   -0.013143   
...                                ...                         ...   
20428                        -0.288535                   -1.216727   
20429                        -0.844466                   -0.692044   
20430                        -0.923885                   -1.143171   
20431                        -0.844466                   -1.055136   
20432                        -1.003304                   -0.780606   

       standardized_bedrooms_per_household  
0                                -0.153863  
1                                -0.262936  
2                                -0.049604  
3                                -0.050417  
4                                -0.033568  
...                                    ...  
20428                             0.076185  
20429                             0.459421  
20430                             0.048373  
20431                             0.157233  
20432                             0.136952  

[20433 rows x 3 columns]